app/gui/src/project-view/stores/project/executionContext.ts

```diff
@@ -487,25 +495,55 @@ export class ExecutionContext extends ObservableV2<ExecutionContextNotification>
           .map((result) => (result.status === 'rejected' ? result.reason : null))
           .filter(isSome)
         if (errors.length > 0) {
-          console.error('Failed to synchronize visualizations:', errors)
+          const result = Err(`Failed to synchronize visualizations: ${errors}`)
+          result.error.log()
+          return result
+        }
+        return Ok()
+      }
+
+      const handleError = (error: ResultError): ExecutionContextState => {
+        // If error tells us that the execution context is missing, we schedule
+        // another sync to re-create it, and set proper state.
+        if (
+          error.payload instanceof LsRpcError &&
+          error.payload.cause instanceof RemoteRpcError &&
+          error.payload.cause.code === ErrorCode.CONTEXT_NOT_FOUND
+        ) {
+          this.sync()
+          return { status: 'not-created' }
+        } else {
+          return newState
         }
       }
 
-      const createResult = await create()
-      if (!createResult.ok) return newState
-      const syncStackResult = await syncStack()
-      if (!syncStackResult.ok) return newState
-      const syncEnvResult = await syncEnvironment()
-      if (!syncEnvResult.ok) return newState
-      this.emit('newVisualizationConfiguration', [new Set(this.visualizationConfigs.keys())])
-      await syncVisualizations()
-      this.emit('visualizationsConfigured', [
-        new Set(state.status === 'created' ? state.visualizations.keys() : []),
-      ])
-      if (this.syncStatus === SyncStatus.SYNCING) {
-        this.syncStatus = SyncStatus.SYNCED
+      this.syncStatus = SyncStatus.CREATING
+      try {
+        if (this.abort.signal.aborted) return newState
+        const createResult = await create()
+        if (!createResult.ok) return newState
+        this.syncStatus = SyncStatus.SYNCING
```

Comment:
I know this state kinda mimicks what was already there, but I think the status logic is still problematic. The `sync` method can be called at any point (because of `pushItem` being callable through external API), and it will queue another sync if the state is not `QUEUED` (and also change the state to `QUEUED` again, masking the ongoing `CREATING`/`SYNCING` state). That means there might be multiple sync promises running in parallel, slightly out of phase. I believe that throws off a lot of assumptions that this code makes.